lib/src/insertion_order_parser.dart

```diff
@@ -0,0 +1,199 @@
+    if (map == null || map.isEmpty) {
+      return InsertionOrder_InsertionOrderBudget();
+    }
+
+    final budget = InsertionOrder_InsertionOrderBudget()
+      ..budgetUnit = _createBudgetUnit(map['budgetUnit'])
+      ..automationType = _createAutomationType(map['automationType']);
+
+    for (Map segmentMap in map['budgetSegments'] ?? []) {
+      budget.budgetSegments.add(_createBudgetSegment(segmentMap));
+    }
+    return budget;
+  }
+
+  /// Creates an
+  /// [InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment]
+  /// instance from [map].
+  static InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment
+      _createBudgetSegment(Map<String, dynamic> map) {
+    if (map == null || map.isEmpty) {
+      InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment();
+    }
+
+    final budgetSegment =
+        InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment()
+          ..budgetAmountMicros = map['budgetAmountMicros'] ?? _emptyEntry
+          ..description = map['description'] ?? _emptyEntry
+          ..campaignBudgetId = map['campaignBudgetId'] ?? _emptyEntry
+          ..dateRange = _createDateRange(map['dateRange']);
+
+    return budgetSegment;
+  }
+
+  /// Creates an
+  /// [InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment_DateRange]
+  /// instance from [map].
+  static InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment_DateRange
+      _createDateRange(Map<String, dynamic> map) {
+    if (map == null || map.isEmpty) {
+      return InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment_DateRange();
+    }
+
+    final dateRange =
+        InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment_DateRange()
+          ..startDate = _createDate(map['startDate'])
+          ..endDate = _createDate(map['endDate']);
+
+    return dateRange;
+  }
+
+  /// Creates an
+  /// [InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment_DateRange_Date]
+  /// instance from [map].
+  static InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment_DateRange_Date
+      _createDate(Map<String, dynamic> map) {
+    if (map == null || map.isEmpty) {
+      return InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment_DateRange_Date();
+    }
+
+    final date =
+        InsertionOrder_InsertionOrderBudget_InsertionOrderBudgetSegment_DateRange_Date()
+          ..year = map['year'] ?? _emptyEntry
+          ..month = map['month'] ?? _emptyEntry
+          ..day = map['day'] ?? _emptyEntry;
+
+    return date;
+  }
+
+  /// Creates an [InsertionOrder_EntityStatus] enum from [map].
+  ///
+  /// Returns default value [ENTITY_STATUS_UNSPECIFIED]
+  /// if [target] doesn't correspond to any enum values.
+  static InsertionOrder_EntityStatus _createEntityStatus(String target) {
+    if (_entityStatusMap.containsKey(target)) return _entityStatusMap[target];
+    return InsertionOrder_EntityStatus.ENTITY_STATUS_UNSPECIFIED;
```

Comment:
enums have a default value already (whichever has a tag of 0). Can we take advantage of the built-in behavior or do we have to set the default ourselves. The advantage of using the default behavior is .hasX() will return false if returning a value by default.